pkce.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2014 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package oauth2
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"io"
+	"net/url"
+)
+
+// PKCEParams holds a PKCE challenge and verifier as described in RFC 7636
+// https://datatracker.ietf.org/doc/html/rfc7636
+type PKCEParams struct {
+	Challenge       string
+	ChallengeMethod string
+	Verifier        string
+}
+
+const (
+	codeChallengeKey       = "code_challenge"
+	codeChallengeMethodKey = "code_challenge_method"
+	codeVerifierKey        = "code_verifier"
+)
+
+// challengeOption should be passed to Config.AuthCodeURL. The option returned sets the code_challenge and code_challenge_method parameters.
+func (p *PKCEParams) challengeOption() AuthCodeOption {
+	return set2Values{k1: codeChallengeKey, v1: p.Challenge, k2: codeChallengeMethodKey, v2: p.ChallengeMethod}
+}
+
+type set2Values struct{ k1, v1, k2, v2 string }
+
+func (p set2Values) setValue(m url.Values) {
+	m.Set(p.k1, p.v1)
+	m.Set(p.k2, p.v2)
+}
+
+// verifierOption should be passed to Config.Exchange. The option returned sets the code_verifier parameters.
+func (p *PKCEParams) verifierOption() AuthCodeOption {
+	return SetAuthURLParam(codeVerifierKey, p.Verifier)
+}
+
+// GeneratePKCEParams generates a code verifier with 32 octets of randomness and a S256 challenge (this follows recommendations in RFC 7636).
+//
+// A fresh verifier should be generated for each AuthCodeURL call.
+func GeneratePKCEParams() *PKCEParams {
+	// "RECOMMENDED that the output of a suitable random number generator be used to create a 32-octet
+	// sequence.  The octet sequence is then base64url-encoded to produce a 43-octet URL-safe string to use as the code verifier."
+	// https://datatracker.ietf.org/doc/html/rfc7636#section-4.1
+	data := make([]byte, 32)
+	if _, err := io.ReadFull(rand.Reader, data); err != nil {
+		panic(err)
+	}
+	verifier := base64.URLEncoding.EncodeToString(data)
+	sha := sha256.Sum256([]byte(verifier))
+	challenge := base64.URLEncoding.EncodeToString(sha[:])
```

Comment:
RFC requires the no padding version of base64 url encoding.

https://datatracker.ietf.org/doc/html/rfc7636#section-3